s3filepath/s3filepath.go

```diff
@@ -0,0 +1,137 @@
+package s3filepath
+
+import (
+	"fmt"
+	"log"
+	"sort"
+
+	"regexp"
+	"time"
+
+	"github.com/mitchellh/goamz/s3"
+)
+
+var (
+	// currently assumes no unix file created timestamp
+	s3Regex = regexp.MustCompile(".*_.*_(.*).json.gz")
+)
+
+// Bucketer interface is useful for testing and showing that
+// we only use List from the goamz library
+type Bucketer interface {
+	List(prefix, delim, marker string, max int) (result *s3.ListResp, err error)
+	Name() string
+	Region() string
+	AccessID() string
+	SecretKey() string
+}
+
+// S3Bucket is our subset of the s3.Bucket class, useful for testing mostly
+type S3Bucket struct {
+	C s3.Bucket
+	// info that makes more sense here
+	BucketName      string
+	BucketRegion    string
+	BucketAccessID  string
+	BucketSecretKey string
+}
+
+// All of these simple accessors are to adjust to Bucketer for testing
+
+// Name returns the name
+func (b *S3Bucket) Name() string { return b.BucketName }
+
+// Region returns the name
+func (b *S3Bucket) Region() string { return b.BucketRegion }
+
+// AccessID returns the name
+func (b *S3Bucket) AccessID() string { return b.BucketAccessID }
+
+// SecretKey returns the name
+func (b *S3Bucket) SecretKey() string { return b.BucketSecretKey }
+
+// List calls the underlying s3.Bucket List method
+func (b *S3Bucket) List(prefix, delim, marker string, max int) (result *s3.ListResp, err error) {
+	return b.C.List(prefix, delim, marker, max)
+}
+
+// S3File holds everything needed to run a COPY on the file
+type S3File struct {
+	// info on which file to get
+	Bucket    Bucketer
+	Schema    string
+	Table     string
+	JSONPaths string
+	Suffix    string
+	DataDate  time.Time
+	ConfFile  string
+}
+
+// GetDataFilename returns the s3 filepath associated with an S3File
+// useful for redshift COPY commands, amongst other things
+func (f *S3File) GetDataFilename() string {
+	return fmt.Sprintf("s3://%s/%s_%s_%s.%s", f.Bucket.Name(), f.Schema, f.Table, f.DataDate.Format(time.RFC3339), f.Suffix)
+}
+
+// We need to sort by data timestamp to find the most recent s3 file
+type byTimeStampDesc []s3.Key
+
+func (k byTimeStampDesc) Len() int           { return len(k) }
+func (k byTimeStampDesc) Swap(i, j int)      { k[i], k[j] = k[j], k[i] }
+func (k byTimeStampDesc) Less(i, j int) bool { return k[i].Key > k[j].Key } // reversing so we get sorted by date desc
+
+// FindLatestInputData looks for the most recent file matching the prefix
+// created by <schema>_<table>, using the RFC3999 date in the filename
+func FindLatestInputData(bucket Bucketer, schema, table, suppliedConf string, targetDate *time.Time) (*S3File, error) {
```

Comment:
While I don't have a better suggestion for a name, I think `FindLatestInputData` may be misleading. Splitting this into 2 separate functions, `FindLatestInputData`, and `FindInputDataForDate`, with an internal helper function encapsulating most of the logic, might be more clear.